src/algorithms/policy_gradient/sac.jl

```diff
@@ -0,0 +1,204 @@
+export SACPolicy, SAC_policy_network, create_sac_policy_network
+
+using Random
+using Flux
+using Flux.Losses: mse
+using Distributions: Normal, logpdf
+
+#### SAC models
+struct SAC_policy_network
+    pre::Any
+    mean::Any
+    log_std::Any
+end
+Flux.@functor SAC_policy_network
+(m::SAC_policy_network)(state) = (x = m.pre(state); (m.mean(x), m.log_std(x)))
+
+function create_sac_policy_network(
+    STATE_SIZE,
+    ACTION_SIZE,
+    f_init;
+    hidden_layer1 = 256,
+    hidden_layer2 = 256,
+    log_std_min = -20f0,
+    log_std_max = 2f0,
+)
+    p = SAC_policy_network(
+        Chain(Dense(STATE_SIZE, hidden_layer1, relu), Dense(hidden_layer1, hidden_layer2, relu)),
+        Chain(Dense(hidden_layer2, ACTION_SIZE, initW = f_init)),
+        Chain(Dense(
+            hidden_layer2,
+            ACTION_SIZE,
+            x -> min(max(x, typeof(x)(log_std_min)), typeof(x)(log_std_max)),
+            initW = f_init,
+        )),
+    )
+end
+
+mutable struct SACPolicy{
+    BA<:NeuralNetworkApproximator,
+    BC1<:NeuralNetworkApproximator,
+    BC2<:NeuralNetworkApproximator,
+    P,
+    R<:AbstractRNG,
+} <: AbstractPolicy
+
+    policy::BA
+    qnetwork1::BC1
+    qnetwork2::BC2
+    target_qnetwork1::BC1
+    target_qnetwork2::BC2
+    γ::Float32
+    ρ::Float32
+    α::Float32
+    batch_size::Int
+    start_steps::Int
+    start_policy::P
+    update_after::Int
+    update_every::Int
+    step::Int
+    rng::R
+end
+
+"""
+    SACPolicy(;kwargs...)
+
+# Keyword arguments
+
+- `policy`,
+- `qnetwork1`,
+- `qnetwork2`,
+- `target_qnetwork1`,
+- `target_qnetwork2`,
+- `start_policy`,
+- `γ = 0.99f0`,
+- `ρ = 0.995f0`,
+- `α = 0.2f0`,
+- `batch_size = 32`,
+- `start_steps = 10000`,
+- `update_after = 1000`,
+- `update_every = 50`,
+- `step = 0`,
+- `rng = Random.GLOBAL_RNG`,
+"""
+function SACPolicy(;
+    policy,
+    qnetwork1,
+    qnetwork2,
+    target_qnetwork1,
+    target_qnetwork2,
+    start_policy,
+    γ = 0.99f0,
+    ρ = 0.995f0,
+    α = 0.2f0,
+    batch_size = 32,
+    start_steps = 10000,
+    update_after = 1000,
+    update_every = 50,
+    step = 0,
+    rng = Random.GLOBAL_RNG,
+)
+    copyto!(qnetwork1, target_qnetwork1)  # force sync
+    copyto!(qnetwork2, target_qnetwork2)  # force sync
+    SACPolicy(
+        policy,
+        qnetwork1,
+        qnetwork2,
+        target_qnetwork1,
+        target_qnetwork2,
+        γ,
+        ρ,
+        α,
+        batch_size,
+        start_steps,
+        start_policy,
+        update_after,
+        update_every,
+        step,
+        rng,
+    )
+end
+
+# TODO: handle Training/Testing mode
+# TODO: if testing act deterministic
+function (p::SACPolicy)(env)
+    p.step += 1
+
+    if p.step <= p.start_steps
+        p.start_policy(env)
+    else
+        D = device(p.policy)
+        s = get_state(env)
+        s = Flux.unsqueeze(s, ndims(s) + 1)
+        # trainmode:
+        action = evaluate(p, s)[1][]
```

Comment:
If I understand it correctly, here action is only a scalar, right? If so, better to add a comment here. the `[]` in the end of the line is a little confusing to me.